File: ToDoList/app/src/main/java/com/example/todolist/madapter/ListViewTaskAdapter.java
package com.example.todolist.madapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.Typeface;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.TextAppearanceSpan;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import com.example.todolist.R;
import com.example.todolist.mactivity.MainActivity;
import com.example.todolist.mclass.Task;
import com.example.todolist.mmodel.CategoryModel;

import java.text.SimpleDateFormat;
import java.util.ArrayList;

public class ListViewTaskAdapter extends ArrayAdapter<Task> {

    private final Context context;
    private final int resource;
    private final ArrayList<Task> items;
    private final ArrayList<Task> itemsTemp;
    private String keyWord = "";
    private final SparseBooleanArray selectedItems; //lưu trữ các phần tử được click vào. ví dụ:
    // position: 0 id1 - Ninh
    // position: 1 id2 - Hiển
    // position: 2 id3 - Tân
    // khi được click, ví dụ:
    // id1 - Ninh - click
    // id2 - Hiển - click
    // thì nó sẽ lưu:
    // 0, true
    // 1, true


    static class ViewHolder {
        ImageView imgIcon;
        RatingBar ratingTask;
        TextView tvTaskName, tvTaskCategory, tvTaskTime;

        public ViewHolder(View view) {
            this.imgIcon = view.findViewById(R.id.imgIcon);
            this.ratingTask = view.findViewById(R.id.ratingTask);
            this.tvTaskName = view.findViewById(R.id.tvTaskName);
            this.tvTaskCategory = view.findViewById(R.id.tvTaskCategory);
            this.tvTaskTime = view.findViewById(R.id.tvTaskTime);
        }
    }

    public ListViewTaskAdapter(@NonNull Context context, int resource, @NonNull ArrayList<Task> objects) {
        super(context, resource, objects);
        this.context = context;
        this.resource = resource;
        this.items = objects;
        this.itemsTemp = new ArrayList<>();
        this.itemsTemp.addAll(objects);
        this.selectedItems = new SparseBooleanArray();
    }

    public void filter(String search) {
        keyWord = search;
        items.clear();
        if (keyWord.length() == 0) {
            items.addAll(itemsTemp);
        } else {
            for (int i = 0; i < itemsTemp.size(); i++) {
                Task currentTask = itemsTemp.get(i);
                if (itemsTemp.get(i).getName().toLowerCase().contains(keyWord.toLowerCase())) {
                    items.add(currentTask);
                }
            }
        }
        notifyDataSetChanged();
    }

    public void checkedStateChanged(int position) {
        if (selectedItems.get(position)) {
            selectedItems.delete(position);
        } else {
            selectedItems.put(position, true);
        }
        notifyDataSetChanged();
    }

    public void destroyActionMode() {
        selectedItems.clear();
        notifyDataSetChanged();
    }

    public void actionItemClicked() {
        for (int i = selectedItems.size() - 1; i >= 0; i--) {
            if (selectedItems.valueAt(i)) {
                Task currentTask = MainActivity.task.get(selectedItems.keyAt(i));
                MainActivity.sqLite.delete(currentTask);
                items.remove(currentTask);
                updateItems(currentTask);
                notifyDataSetChanged();
            }
        }
    }

    public void updateItems(Task currentTask) {
        itemsTemp.remove(currentTask);
    }

    @SuppressLint("SetTextI18n")
    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        View view = convertView;
        ViewHolder viewHolder;

        if (view == null) {
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater.inflate(resource, parent, false);
            viewHolder = new ViewHolder(view);
            view.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) view.getTag();
        }

        Task task = items.get(position);
        if (task != null) {
            viewHolder.tvTaskName.setText(task.getName());
            viewHolder.tvTaskCategory.setText(CategoryModel.getNameByID(task.getType()));
            viewHolder.ratingTask.setRating(task.getRating());
            @SuppressLint("SimpleDateFormat") SimpleDateFormat simpleDateFormat = new SimpleDateFormat("HH:mm");
            viewHolder.tvTaskTime.setText("Starting at: " + simpleDateFormat.format(task.getTime().getTime()));
            viewHolder.imgIcon.setImageResource(task.getIcon());

            if (task.getName().toLowerCase().contains(keyWord)) {
                int posStart = task.getName().toLowerCase().indexOf(keyWord);
                int posEnd = posStart + keyWord.length();

                Spannable spannable = new SpannableString(task.getName());
                TextAppearanceSpan textAppearanceSpan = new TextAppearanceSpan(null, Typeface.ITALIC, -1, ColorStateList.valueOf(Color.BLACK), null);
                spannable.setSpan(textAppearanceSpan, posStart, posEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                viewHolder.tvTaskName.setText(spannable);
            }

            view.setBackgroundResource(android.R.color.transparent);
            if (selectedItems.size() > 0) {
                if (selectedItems.get(position)) {
                    view.setBackgroundColor(Color.parseColor("#a9a9a9"));
                }
            }
        }
        return view;
    }
}

File: ToDoListKotlin/app/src/main/java/com/example/todolistkotlin/activity/adapter/ListViewTaskAdapter.kt
package com.example.todolistkotlin.activity.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Typeface
import android.text.Spannable
import android.text.SpannableString
import android.text.Spanned
import android.text.style.TextAppearanceSpan
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import com.example.todolistkotlin.R
import com.example.todolistkotlin.activity.activity.MainActivity
import com.example.todolistkotlin.activity.classes.Task
import com.example.todolistkotlin.activity.model.CategoryModel
import java.text.SimpleDateFormat
import java.util.*


class ListViewTaskAdapter(
    getContext: Context,
    private val resource: Int,
    objects: ArrayList<Task>
) : ArrayAdapter<Task?>(getContext, resource, objects as List<Task?>) {
    private val items: ArrayList<Task> = objects
    private val itemsTemp: ArrayList<Task> = ArrayList<Task>()
    private var keyWord = ""
    private val selectedItems: SparseBooleanArray//lưu trữ các phần tử được click vào. ví dụ:

    // position: 0 id1 - Ninh
    // position: 1 id2 - Hiển
    // position: 2 id3 - Tân
    // khi được click, ví dụ:
    // id1 - Ninh - click
    // id2 - Hiển - click
    // thì nó sẽ lưu:
    // 0, true
    // 1, true

    internal class ViewHolder(view: View?) {
        var imgIcon: ImageView = view!!.findViewById(R.id.imgIcon)
        lateinit var ratingTask: RatingBar
        lateinit var tvTaskName: TextView
        lateinit var tvTaskCategory: TextView
        lateinit var tvTaskTime: TextView

        init {
            if (view != null) {
                ratingTask = view.findViewById(R.id.ratingTask)
                tvTaskName = view.findViewById(R.id.tvTaskName)
                tvTaskCategory = view.findViewById(R.id.tvTaskCategory)
                tvTaskTime = view.findViewById(R.id.tvTaskTime)
            }

        }
    }

    fun filter(search: String) {
        keyWord = search
        items.clear()
        if (keyWord.isEmpty()) {
            items.addAll(itemsTemp)
        } else {
            for (i in itemsTemp.indices) {
                val currentTask: Task = itemsTemp[i]
                if (itemsTemp[i].getName().toLowerCase(Locale.ROOT)
                        .contains(keyWord.toLowerCase(Locale.ROOT))
                ) {
                    items.add(currentTask)
                }
            }
        }
        notifyDataSetChanged()
    }

    fun checkedStateChanged(position: Int) {
        if (selectedItems[position]) {
            selectedItems.delete(position)
        } else {
            selectedItems.put(position, true)
        }
        notifyDataSetChanged()
    }

    fun destroyActionMode() {
        selectedItems.clear()
        notifyDataSetChanged()
    }

    fun actionItemClicked() {
        for (i in selectedItems.size() - 1 downTo 0) {
            if (selectedItems.valueAt(i)) {
                val currentTask: Task? = MainActivity.task?.get(selectedItems.keyAt(i))
                if (currentTask != null) {
                    MainActivity.sqLite?.delete(currentTask)
                }
                items.remove(currentTask)
                updateItems(currentTask)
                notifyDataSetChanged()
            }
        }
    }

    fun updateItems(currentTask: Task?) {
        itemsTemp.remove(currentTask)
    }

    @SuppressLint("SetTextI18n")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val viewHolder: ViewHolder

        if (view == null) {
            val inflater =
                context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = inflater.inflate(resource, parent, false)
            viewHolder = ViewHolder(view)
            view.tag = viewHolder
        } else {
            viewHolder = view.tag as ViewHolder
        }

        val task: Task = items[position]
        viewHolder.tvTaskName.text = task.getName()
        viewHolder.tvTaskCategory.text = CategoryModel.getNameByID(task.getType())
        viewHolder.ratingTask.rating = task.getRating().toFloat()
        @SuppressLint("SimpleDateFormat") val simpleDateFormat = SimpleDateFormat("HH:mm")
        viewHolder.tvTaskTime.text =
            "Starting at: " + simpleDateFormat.format(task.getTime().time)
        viewHolder.imgIcon.setImageResource(task.getIcon())

        if (task.getName().toLowerCase(Locale.ROOT).contains(keyWord)) {
            val posStart: Int = task.getName().toLowerCase(Locale.ROOT).indexOf(keyWord)
            val posEnd = posStart + keyWord.length
            val spannable: Spannable = SpannableString(task.getName())
            val textAppearanceSpan = TextAppearanceSpan(
                null, Typeface.ITALIC, -1, ColorStateList.valueOf(
                    Color.BLACK
                ), null
            )
            spannable.setSpan(
                textAppearanceSpan,
                posStart,
                posEnd,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            viewHolder.tvTaskName.text = spannable
        }

        view!!.setBackgroundResource(R.color.nude)
        if (selectedItems.size() > 0) {
            if (selectedItems[position]) {
                view.setBackgroundColor(Color.parseColor("#a9a9a9"))
            }
        }
        return view
    }

    init {
        itemsTemp.addAll(objects)
        selectedItems = SparseBooleanArray()
    }
}
